xtokens/src/lib.rs

```diff
@@ -467,43 +469,140 @@ pub mod module {
 				assets.len() <= T::MaxAssetsForTransfer::get(),
 				Error::<T>::TooManyAssetsBeingSent
 			);
+			let origin_location = T::AccountIdToMultiLocation::convert(who.clone());
 
-			// We check that all assets are valid and share the same reserve
-			for i in 0..assets.len() {
+			let mut non_fee_reserve: Option<MultiLocation> = None;
+			let asset_len = assets.len();
+			for i in 0..asset_len {
 				let asset = assets.get(i).ok_or(Error::<T>::AssetIndexNonExistent)?;
 				ensure!(
 					matches!(asset.fun, Fungibility::Fungible(x) if !x.is_zero()),
 					Error::<T>::InvalidAsset
 				);
-				ensure!(
-					fee.reserve() == asset.reserve(),
-					Error::<T>::DistinctReserveForAssetAndFee
-				);
+				// `assets` includes fee, the reserve location is decided by non fee asset
+				if (fee != *asset && non_fee_reserve.is_none()) || asset_len == 1 {
+					non_fee_reserve = asset.reserve();
+				}
+				// make sure all non fee assets share the same reserve
+				if non_fee_reserve.is_some() {
+					ensure!(
+						non_fee_reserve == asset.reserve(),
+						Error::<T>::DistinctReserveForAssetAndFee
+					);
+				}
 			}
 
-			let (transfer_kind, dest, reserve, recipient) = Self::transfer_kind(&fee, &dest)?;
-			let mut msg = match transfer_kind {
-				SelfReserveAsset => Self::transfer_self_reserve_asset(
-					assets.clone(),
-					fee.clone(),
-					dest.clone(),
-					recipient,
-					dest_weight,
-				)?,
-				ToReserve => {
-					Self::transfer_to_reserve(assets.clone(), fee.clone(), dest.clone(), recipient, dest_weight)?
+			// In case that fee reserve != asset reserve, there are two xcm sent from
+			// sender. first xcm sent to fee reserve chain which route to dest chain. second
+			// xcm directly sent to dest chain. the fee amount in fee asset is split into
+			// two parts. 1. fee asset sent to fee reserve chain = fee_amount - min_xcm_fee
+			// 2. fee asset sent to dest reserve chain = min_xcm_fee
+			// the first part + second part = fee amount in fee asset
+			let fee_reserve = fee.reserve();
+			if fee_reserve != non_fee_reserve {
+				// Current only support `ToReserve` with relay-chain asset as fee. other case
+				// like `NonReserve` or `SelfReserve` with relay-chain fee is not support.
+				ensure!(non_fee_reserve == dest.chain_part(), Error::<T>::InvalidAsset);
```

Comment:
should support NonReserve, i.e. Karura-[RMRK, KSM(fee)]->Bifrost

Reply:
No support `NonReserve` for now, user can first transfer RMRK from Karura to Statemine, then transfer RMRK from Statemine to Bifrost.